presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -107,7 +120,59 @@ public void process(Iterable<QueryExecution> queries)
         return ImmutableMap.copyOf(pools);
     }
 
-    private void updateNodes()
+    private MemoryPoolAssignmentsRequest updateAssignments(Iterable<QueryExecution> queries)
+    {
+        ClusterMemoryPool reservedPool = pools.get(RESERVED_POOL);
+        long version = memoryPoolAssignmentsVersion.incrementAndGet();
+        if (reservedPool != null && allAssignmentsHavePropagated(queries)) {
+            if (reservedPool.getQueries() == 0) {
+                QueryExecution biggestQuery = null;
+                long maxMemory = -1;
+                for (QueryExecution queryExecution : queries) {
+                    long bytesUsed = queryExecution.getQueryInfo().getQueryStats().getTotalMemoryReservation().toBytes();
+                    if (bytesUsed > maxMemory) {
+                        biggestQuery = queryExecution;
+                        maxMemory = bytesUsed;
+                    }
+                }
+                for (QueryExecution queryExecution : queries) {
+                    if (queryExecution.getQueryId().equals(biggestQuery.getQueryId())) {
+                        queryExecution.setMemoryPool(new VersionedMemoryPoolId(RESERVED_POOL, version));
```

Comment:
I thought we only wanted to do this if the general pool was blocked.  Also won't this move the query in the reserved pool back to the general pool if another bigger query comes along?